[SMTChecker] Support compound shifts and bitwise and, or, and xor

mijovic opened the pull request:
Implements what is missing from #9043

mijovic committed:
[SMTChecker] Implementing compound bitwise And/Or/Xor operators

mijovic committed:
[SMTChecker] Updating old and adding new tests for compound bitwise and operator.

mijovic committed:
[SMTChecker] Updating old and adding new tests for compound bitwise xor operator.

mijovic committed:
[SMTChecker] Updating old and adding new tests for compound bitwise or operator.

mijovic committed:
[SMTChecker] Supporting compound shift operators.

mijovic committed:
[SMTChecker] Updating old and adding new tests for compound shift operators.

leonardoalt commented on Changelog.md:
```diff
@@ -8,6 +8,7 @@ Compiler Features:
  * SMTChecker: Support events and low-level logs.
  * SMTChecker: Support ``revert()``.
  * SMTChecker: Support shifts.
+ * SMTChecker: Support compound and, or, and xor operators.
```
And shifts?

mijovic replied:
I was thinking that it is not needed, as in the same release we added support for shifts. Will add one as probably it is still fine to have both shifts and compound shifts in same release.

leonardoalt replied:
Ah ok, makes sense

mijovic replied:
Ok, let's than add changelog item if PR doesn't make it to this release?

leonardoalt replied:
Yep!

leonardoalt commented on libsolidity/formal/SMTEncoder.h:
```diff
@@ -119,6 +119,15 @@ class SMTEncoder: public ASTConstVisitor
+		Expression const& _right,
+		TypePointer const& _commonType
+	);
+
```
Extra blank

leonardoalt commented on libsolidity/formal/SMTEncoder.cpp:
```diff
@@ -357,25 +357,17 @@ void SMTEncoder::endVisit(Assignment const& _assignment)
-			"Assertion checker does not yet implement this assignment operator."
-		);
-	}
+		solAssert(false, "");
```
Now that the `set` above is completed with all the assignment operators, we could replace the section by
```
solAssert(TokenTraits::isAssignmentOp(op), "");
```

leonardoalt commented on libsolidity/formal/SMTEncoder.cpp:
```diff
@@ -1597,8 +1610,25 @@ smtutil::Expression SMTEncoder::compoundAssignment(Assignment const& _assignment
+	if (compoundToBitwise.count(op))
+		return bitwiseOperation(
+			compoundToBitwise.at(op),
+			_assignment.leftHandSide(),
```
```suggestion
                    decl ? currentValue(*decl) : expr(_assignment.leftHandSide()),
```
otherwise the assignment won't be propagated to the actual variable besides the identifier

leonardoalt replied:
Hm the tests pass so now I'm wondering why it works without this change

leonardoalt replied:
I think it works without this suggestion because of some recent changes I made on the `Identifier` visit.

mijovic replied:
Fixed it now in any case

leonardoalt reviewed (commented):
Some minor changes, but looks good overall!

leonardoalt commented on test/libsolidity/smtCheckerTests/operators/compound_bitwise_and_uint.sol:
```diff
@@ -0,0 +1,34 @@
+
+contract C {
+	function f(bool b) public pure {
+		uint v = 1;
```
Indentation is off.

leonardoalt commented on test/libsolidity/smtCheckerTests/operators/compound_bitwise_and_uint.sol:
```diff
@@ -0,0 +1,34 @@
+        uint16 y = 0xffff;
+        y &= x;
+        assert(y == 0xff);
+        assert(y == 0xffff);
```
I guess this one fails?

mijovic replied:
Yes, should have added comment as well. Adding one now :)

leonardoalt commented on test/libsolidity/smtCheckerTests/operators/compound_bitwise_or_uint.sol:
```diff
@@ -0,0 +1,29 @@
+
+contract C {
+	function f(bool b) public pure {
+		uint v = b ? 7 : 3;
```
Indentation is off.

leonardoalt commented on test/libsolidity/smtCheckerTests/operators/compound_bitwise_or_int_1.sol:
```diff
@@ -7,4 +7,3 @@ contract C {
 	}
```
Can you add an assertion here to test that it's also working for arrays?

leonardoalt commented on test/libsolidity/smtCheckerTests/operators/compound_bitwise_or_uint_1.sol:
```diff
@@ -7,4 +7,3 @@ contract C {
 	}
```
Also here

leonardoalt commented on test/libsolidity/smtCheckerTests/operators/compound_bitwise_or_uint_2.sol:
```diff
@@ -10,4 +10,3 @@ contract C {
 	}
```
Also here

leonardoalt commented on test/libsolidity/smtCheckerTests/operators/compound_bitwise_or_uint_3.sol:
```diff
@@ -10,4 +10,3 @@ contract C {
 	}
```
Also here

leonardoalt commented on test/libsolidity/smtCheckerTests/operators/compound_bitwise_xor_uint.sol:
```diff
@@ -0,0 +1,29 @@
+
+contract C {
+	function f(bool b) public pure {
+		uint v = b ? 7 : 3;
```
Indentation off

leonardoalt commented on test/libsolidity/smtCheckerTests/operators/shifts/compound_shift_left.sol:
```diff
@@ -0,0 +1,30 @@
+        a <<= b;
+        return a;
+    }
+	function t() public pure {
```
Indentation

leonardoalt commented on test/libsolidity/smtCheckerTests/operators/shifts/compound_shift_right.sol:
```diff
@@ -0,0 +1,30 @@
+        a >>= b;
+        return a;
+    }
+	function t() public pure {
```
Indentation

mijovic commented:
@leonardoalt Updated

mijovic commented on libsolidity/formal/SMTEncoder.cpp:
```diff
@@ -394,9 +373,9 @@ void SMTEncoder::endVisit(Assignment const& _assignment)
-				expr(_assignment.rightHandSide(), type);
+			auto rightHandSide = op == Token::Assign ?
+				expr(_assignment.rightHandSide(), type) :
+				compoundAssignment(_assignment);
```
After all updates, only place where `compoundOps` was used was here, so I removed variable and updated this ternary with appropriate condition.

leonardoalt commented on test/libsolidity/smtCheckerTests/operators/shifts/compound_shift_right.sol:
```diff
@@ -0,0 +1,30 @@
+
+        assert(f(0x4266, 0x11) == 0);
+        // Fails because the above is true.
+        assert(f(0x4266, 0x11) == 1);
```
@mijovic the cli tests are taking too long. I suspect it's because of the failing asserts in this example and the respective `left`. Failing asserts lead to counterexample generation which can take the solver a really long time, especially for more complicated cases. This doesn't necessarily look like such a case, but it actually is because it converts back and forth between SMT Integers and BitVectors, so that complicates the counterexamples as well. I'd suggest maybe removing this assertion and the similar one in the `left` case so we can see whether the cli time decreases.

mijovic replied:
I did remove those checks, let's hope that it will pass now

leonardoalt commented:
@mijovic still the same =/ maybe remove the last asserts as well?

mijovic commented:
> @mijovic still the same =/ maybe remove the last asserts as well?

Locally it looks like failing assertions for arrays are making it slow, so I removed those. Also, removed ternary from `_uint` tests. Should pass now

leonardoalt commented on test/libsolidity/smtCheckerTests/operators/compound_bitwise_and_uint.sol:
```diff
@@ -0,0 +1,32 @@
+}
+// ----
+// DeclarationError 7576: (150-151): Undeclared identifier.
+// DeclarationError 7576: (175-176): Undeclared identifier.
```
Compilation didn't go through here

mijovic replied:
I need coffe :)

leonardoalt commented on test/libsolidity/smtCheckerTests/operators/compound_bitwise_or_int_1.sol:
```diff
@@ -2,9 +2,10 @@ pragma experimental SMTChecker;
 contract C {
 	int[1][20] c;
 	function f(bool b) public {
+	    require(c[10][0] == 0);
```
Indentation
